File: AsyncFSM.py
import asyncio
from enum import Enum
import AsyncEvent
#The startStateFetcher should be a callable returning initial state
#This is done to suppport the state machines which have multiple entry points, so that
#the callable can point to the appropriate starting point depending on the conditions
#in which the state machine was started

class AFSMException(Exception):
    def __init__(self, *args: object) -> None:
        super().__init__(*args)

class FinalityReachedException(AFSMException):
    def __init__(self):
        super().__init__("State machine can't accept events anymore as it has already reached finality")

class EventErrors(Enum):
    unhandledEvent = 'unhandled'
    conditionalReject = 'conditionalReject'

class SpecialEvents(Enum):
    defer = 'defer'

class AFSM:
    def __init__(self, startStateFetcher):
        self.currState = LauncherState(startStateFetcher())
        self.prevState = None
        self.endingEvent = asyncio.Event()
        self.endingNotifier = AsyncEvent.AsyncEvent()
        self.eventQueue = []
        self.deferalQueue = []
        self.conclusion = None

    async def start(self):
        await self.handleEvent("launch")

    async def startAndAwaitEnd(self):
        await self.handleEvent("launch")
        await self.endingEvent.wait()

    async def findNextState(self, evt, *args):
        nextState = await self.currState.react(evt, *args)
        if EventErrors.unhandledEvent == nextState:
            if issubclass(type(self.currState), AFSM):
                try:
                    nextState = await self.currState.handleEvent(evt, *args)
                except FinalityReachedException:
                    pass
            if EventErrors.unhandledEvent != nextState:
                nextState = None
        return nextState

    async def handleStateEntry(self, state):
        res = await state.after_entry()
        childStateMachine = (state  
                             if issubclass(type(state), AFSM) else
                             None)
        if childStateMachine is None:
            return res
        else:
            return await childStateMachine.start()

    async def handleStateExit(self, state):
        childStateMachine = (state  
                            if issubclass(type(state), AFSM) else
                            None)
        if childStateMachine is not None:
            await self.handleStateExit(childStateMachine.currState)
        await state.before_exit()

    async def processEvent(self, evt, *args):
        nextState = await self.findNextState(evt, *args)
        if EventErrors.unhandledEvent == nextState:
            await self.onUnconsumedEvt(evt)
        elif EventErrors.conditionalReject == nextState:
            await self.onConditionalReject(evt)
        elif SpecialEvents.defer == nextState:
            self.deferalQueue.append([evt, args])
        elif nextState != None:
            await self.handleStateExit(self.currState)
            self.prevState = self.currState
            self.currState = nextState
            endResult = await self.handleStateEntry(self.currState)
            if self.currState.isFinal:
                self.conclusion = endResult
                await self.endingNotifier(self.conclusion)
                self.endingEvent.set()
            else:
                await self.processDeferalQueueLoop()

    async def handleEvent(self, evt, *args):
        if(self.currState.isFinal):
            raise FinalityReachedException()
        self.eventQueue.append([evt, args])
        if 1 == len(self.eventQueue):
            await self.processEvtQueueLoop()

    async def processDeferalQueueLoop(self):
        localQueue = []
        localQueue, self.deferalQueue = self.deferalQueue, localQueue
        while 0 < len(localQueue):
            nextEvt = localQueue[0]
            await self.processEvent(nextEvt[0], *nextEvt[1])
            localQueue.pop(0)

    async def processEvtQueueLoop(self):
        while 0 < len(self.eventQueue):
            nextEvt = self.eventQueue[0]
            await self.processEvent(nextEvt[0], *nextEvt[1])
            self.eventQueue.pop(0)

    async def registerForEnd(self, callback):
        self.endingNotifier += callback

    async def onUnconsumedEvt(self, evt):
        pass

    async def onConditionalReject(self, ev2t):
        pass


class AFSMState:
    def __init__(self, isFinal):
        self.isFinal = isFinal

    async def after_entry(self):
        pass

    async def before_exit(self):
        pass

    async def react(self, evt, *args):
        funcName = "on_" + evt
        if hasattr(self, funcName):
            return await getattr(self, funcName)(*args)
        else:
            return EventErrors.unhandledEvent

class CompositeState(AFSMState, AFSM):
    def __init__(self, startStateFetcher):
        AFSMState.__init__(self, False)
        AFSM.__init__(self, startStateFetcher)

class LauncherState(AFSMState):
    def __init__(self, initialState):
        super().__init__(False)
        self.initialState = initialState

    async def on_launch(self):
        return self.initialState

File: AsyncFSM.js
class AFSMError extends Error{
    constructor(message){
        super(message)
        this.name = this.constructor.name 
    }
}

class FinalityReachedException extends AFSMError
{
    constructor(){
        super("State machine has reched final state and can't process any new events")
    }
}

class SMInactiveException extends AFSMError
{
    constructor(){
	    super("State machine needs to be started by calling the start() method")
        this.name = "SMInactiveException"
    }
}

class UnhandledEvtException extends AFSMError
{
    constructor(stateName, evtName){
	    super(`Event: ${evtName} is unhandled in state: ${stateName}`)
    }
}

class ImproperReactionException extends AFSMError
{
    constructor(stateName, evtName, reactionType){
	    super(`Improper reaction from state: ${stateName}, while handling event: ${evtName}, the reaction should be either a new state or a member of "SpecialTransition", but the type of the reaction was of type: ${reactionType}`)
    }
}

class RecursiveEventException extends AFSMError
{
    constructor(){
	    super(`Raising and event on FSM while it is already processing an event`)
    }
}

const SpecialTransition = 
{
	nullTransition : "nullTransition",
	deferralTransition : "deferralTransition"
};

class State
{
	constructor(isFinal = false){
        this.isFinal = isFinal
        this.name = this.constructor.name
    }

    on_launch() { return SpecialTransition.nullTransition }
	onEntry() { }
	beforeExit() {}
	final(){ return this.isFinal }
    react(evtName, evtData)
    {
        let expectedEvtHandlerMethodName = "on_" + evtName
        if(this[expectedEvtHandlerMethodName] == undefined)
            throw new UnhandledEvtException(this.name, evtName)

        let transition = null
        if(evtData == null){
            transition = (this[expectedEvtHandlerMethodName])()
        }
        else{
            transition = (this[expectedEvtHandlerMethodName])(evtData)
        }
        
        if (transition instanceof State){
            this.beforeExit()
            return transition
        }
        else if (SpecialTransition[transition] != undefined){
            return transition
        }
        else{
            throw new ImproperReactionException(this.name, evtName, typeof transition)
        }
    }
}

class FSM
{
	constructor(startStateFetcher, logger = (message)=>{ console.log(message) } )
	{
        this.currState = startStateFetcher()
        this.logger = logger
        this.started = false
        this.smBusy = false//FSM is bust processing an evt
        this.deferralQueue = []
    }

    checkIfFSMReadyToHandleEvt(){
        if (!this.started)
            throw new SMInactiveException()
        else if (this.currState.final())
            throw new FinalityReachedException()
        else if(this.smBusy)
            throw new RecursiveEventException()
    }

	handleEvent(evtName, evtData = null)
	{
        this.checkIfFSMReadyToHandleEvt()
        this.processSingleEvent(evtName, evtData)
	}

    processSingleEvent(evtName, evtData){
        this.smBusy = true
        let transition = null
        try{
            transition = this.currState.react(evtName, evtData)
        }
        finally{
            this.smBusy = false
        }

        this.smBusy = false
        if(transition instanceof State){
            this.currState = transition
            this.handleStateEntry(this.currState)
        }
        else if(SpecialTransition.deferralTransition == transition){
            this.deferralQueue.push([evtName, evtData])
        }
    }

	start()
	{
		this.started = true
        if (this.currState.final())
            throw new FinalityReachedException()
        this.handleStateEntry(this.currState)
	}

	processDeferralQueue(){
        if (0 == this.deferralQueue.length){
            return
        }

		let local = []
        let temp = local
        local = this.deferralQueue
        this.deferralQueue = temp
		
		while (0 < local.length){
            try{
                this.checkIfFSMReadyToHandleEvt()
                let [evtName, evtData] = local[0]
                this.processSingleEvent(evtName, evtData)
            }
            catch(err){
                this.logger(`Error while processing deferral queue: ${err.message}`)
            }
            finally{
                local.pop()
            }
		}
	}
	
	handleStateEntry(state){
		state.onEntry()
        this.handleEvent("launch")
		this.processDeferralQueue()
	}
};

class CompositeState extends State
{
	constructor(startStateFetcher, 
                evtProcessorDictionary,
                 isFinal = false)
	{
        super(evtProcessorDictionary, isFinal)
        this.fsm = new FSM(startStateFetcher)
        this.fsm.start()
    }

    react(name, evtData)
    {
        try{
            transition = super.react(name, evtData)
            if(transition instanceof State){
                this.fsm.currState.beforeExit()
                return transition
            }
        }
        catch(err){
            if(!(err instanceof UnhandledEvtException)){
                throw err
            }
        }

        try{
            this.fsm.handleEvent(evt)
        }
        catch(err){
            if(!(err instanceof FinalityReachedException)){
                throw err
            }
        }
        return SpecialTransition.nullTransition
    }
}

module.exports.FSM = FSM
module.exports.State = State
module.exports.CompositeState = CompositeState
module.exports.FinalityReachedException = FinalityReachedException
module.exports.SMInactiveException = SMInactiveException
module.exports.UnhandledEvtException = UnhandledEvtException
module.exports.ImproperReactionException = ImproperReactionException
module.exports.SpecialTransition = SpecialTransition